crates/runtime/src/derive/serve.rs

```diff
@@ -74,58 +81,57 @@ async fn serve_session<L: LogHandler>(
     runtime: &Runtime<L>,
     shape: &mut doc::Shape,
 ) -> anyhow::Result<Option<Request>> {
-    recv_client_open(&mut open, &db)?;
+    recv_client_open(&mut open, &db).await?;
 
     // Start connector stream and read Opened.
     let (mut connector_tx, mut connector_rx) = connector::start(runtime, open.clone()).await?;
     let opened = TryStreamExt::try_next(&mut connector_rx).await?;
 
     let (task, mut validators, mut accumulator, mut last_checkpoint, opened) =
-        recv_connector_opened(&db, &open, opened)?;
+        recv_connector_opened(&db, &open, opened).await?;
 
     () = co.yield_(opened).await;
 
     let mut buf = bytes::BytesMut::new();
     loop {
-        // Loop over EOF, Reset, and Open until an initial Read or Flush.
-        let initial: Request = loop {
-            match request_rx.try_next().await? {
-                None => {
-                    drain_connector(connector_tx, connector_rx).await?;
-                    return Ok(None);
-                }
-                Some(reset @ Request { reset: Some(_), .. }) => {
-                    connector_tx.try_send(reset).expect("sender is empty");
-                }
-                Some(open @ Request { open: Some(_), .. }) => {
-                    drain_connector(connector_tx, connector_rx).await?;
-                    return Ok(Some(open));
-                }
-                Some(read @ Request { read: Some(_), .. }) => break read,
-                Some(flush @ Request { flush: Some(_), .. }) => break flush,
-                request => {
-                    return verify("client", "EOF, Reset, Open, Read, or Flush").fail(request)
-                }
-            }
-        };
-
-        let mut txn = Transaction::new();
-        txn.started_at = std::time::SystemTime::now();
-
-        enum Step {
-            ClientRx(Option<Request>),
-            ConnectorRx(Option<Response>),
-            ConnectorTx(Result<(), mpsc::SendError>),
-        }
         let mut saw_flush = false;
         let mut saw_flushed = false;
         let mut saw_reset = false;
         let mut send_fut = None;
-        let mut step = Step::ClientRx(Some(initial));
+        let mut txn = Transaction::new();
 
         // Loop over client requests and connector responses until the transaction has flushed.
-        loop {
+        while !saw_flushed {
+            enum Step {
+                ClientRx(Option<Request>),
+                ConnectorRx(Option<Response>),
+                ConnectorTx(Result<(), mpsc::SendError>),
+            }
+
+            let step = if let Some(forward) = &mut send_fut {
+                tokio::select! {
+                    result = forward => Step::ConnectorTx(result),
+                    response = connector_rx.try_next() => Step::ConnectorRx(response?),
+                }
+            } else {
+                tokio::select! {
+                    request = request_rx.try_next(), if !saw_flush => Step::ClientRx(request?),
```

Comment:
I'm not quite following the question. The role of !saw_flush is to disable this branch once `saw_flush` is true, so that we stop receiving from `request_rx` (but are still listening to `connector_rx`.

Panics only happen if no patterns are matched once all futures complete, and for that to be the case there needs to be a left-hand-side pattern that could potentially not match (like `Some(response) = connector_rx.try_next()....`). That's deliberately not done here, and we handle Some/None elsewhere.

As compared to the journal reads returning None case for preview, which _did_ use a `Some(read) = ` pattern construction because it wasn't ever okay for it to return None.